api/java/java_download.go

```diff
@@ -2,51 +2,104 @@ package java
-	r.HandleFunc("/java/{group:.+}/{artifact:.+}/{version:.+}/{filename:[^/]+}{type:\\.jar|\\.pom}{checksum:\\.md5|\\.sha1}", je.javaDownloadArtifactChecksumRouter)
+func (je endpoints) AppendEndpoints(r *mux.Router, handlers ...func(w http.ResponseWriter, r *http.Request)) {
+
+	var javaDarFunc func(w http.ResponseWriter, r *http.Request)
+	var javaDacrFunc func(w http.ResponseWriter, r *http.Request)
+
+	if handlers == nil {
+		// If we don't submit any options, we should go for the default functions to handle these endpoints.
+		javaDarFunc = je.javaDownloadArtifactRouter
+		javaDacrFunc = je.javaDownloadArtifactChecksumRouter
+	} else {
+		// However, for perhaps test purposes - if we want to override these handlers then we should.
+		javaDarFunc = handlers[0]
+		javaDacrFunc = handlers[1]
+	}
+
+	r.HandleFunc("/java/{group:.+}/{artifact:.+}/{version:.+}/{filename:[^/]+}{type:\\.jar|\\.pom}", javaDarFunc)
+	r.HandleFunc("/java/{group:.+}/{artifact:.+}/{version:.+}/{filename:[^/]+}{type:\\.jar|\\.pom}{checksumType:\\.md5|\\.sha1}", javaDacrFunc)
 }
 
 // For checksums, we want to attempt to download something from the proxy - but if it doesn't exist, it's probably
 // better if we generate a checksum using the artifact we've got as some files don't have a checksum.
 // TODO: Maybe we make this choice configurable?
-func (je Endpoints) javaDownloadArtifactChecksumRouter(w http.ResponseWriter, r *http.Request) {
+func (je endpoints) javaDownloadArtifactChecksumRouter(w http.ResponseWriter, r *http.Request) {
 
 	ja := RequestToArtifact(mux.Vars(r))
 	log.WithFields(log.Fields{"module": "api", "action": "downloadChecksum"}).Info(ja.ToString())
 	je.javaProxiedArtifactAction(w, r, ja)
 }
 
-func (je Endpoints) javaDownloadArtifactRouter(w http.ResponseWriter, r *http.Request) {
+func (je endpoints) javaDownloadArtifactRouter(w http.ResponseWriter, r *http.Request) {
 
 	ja := RequestToArtifact(mux.Vars(r))
 	log.WithFields(log.Fields{"module": "api", "action": "downloadArtifact"}).Info(ja.ToString())
 	je.javaProxiedArtifactAction(w, r, ja)
 }
 
-func (je Endpoints) javaProxiedArtifactAction(w http.ResponseWriter, r *http.Request, ja Artifact) {
+func (je endpoints) javaProxiedArtifactAction(w http.ResponseWriter, r *http.Request, ja Artifact) {
 
-	// If file exists in index - attempt to serve,
+	// If the file does not exist in the index and cannot be served from within the system...
 	if !je.ArtifactIndex.IsDownloadedArtifact(ja.GetIdentifier(), ja.Type) {
 
-		mavenCentral := "https://repo.maven.apache.org/maven2/" + strings.Replace(ja.Group, ".", "/", -1) + "/" + ja.Artifact + "/" + ja.Version + "/" + ja.Filename
-		je.ArtifactStorage.DownloadArtifactToStorage(mavenCentral, ja.GetStorageIdentifier())
+		// Cycle through the repositories that are available.
+		proxies := viper.GetStringSlice("java.proxies")
+		for _, proxy := range proxies {
+
+			u, _ := url.Parse(proxy)
```

Comment:
Do we we do any input sanitising to ensure that the proxy URLs passed in are valid URLs to justify throwing any errors away here?

Reply:
tbh, in this case I think I need to do more than that. This slice of proxies won't work as intended unless I check to see if there has been a download, and if there hasn't move onto the next proxy AND break if we do get a successful download.

However, to prevent this from dragging on, I've raised #16